src/lib/sql/sql_identifier_context.cpp

```diff
@@ -0,0 +1,84 @@
+#include "sql_identifier_context.hpp"
+
+#include "utils/assert.hpp"
+#include "sql_identifier_context_proxy.hpp"
+
+using namespace std::string_literals;  // NOLINT
+
+namespace opossum {
+
+void SQLIdentifierContext::set_column_name(const std::shared_ptr<AbstractExpression>& expression, const std::string& column_name) {
+  auto& entry = _find_or_create_expression_entry(expression);
+  if (entry.identifier) {
+    entry.identifier->column_name = column_name;
+  } else {
+    entry.identifier.emplace(column_name);
+  }
+}
+
+void SQLIdentifierContext::set_table_name(const std::shared_ptr<AbstractExpression>& expression, const std::string& table_name) {
+  auto& entry = _find_or_create_expression_entry(expression);
+  if (!entry.identifier) entry.identifier.emplace(expression->as_column_name());
+
+  entry.identifier->table_name = table_name;
+}
+
+std::shared_ptr<AbstractExpression> SQLIdentifierContext::resolve_identifier_relaxed(const SQLIdentifier& identifier) const {
+  std::vector<std::shared_ptr<AbstractExpression>> matching_expressions;
+  for (const auto& entry : _entries) {
+    if (!entry.identifier) continue;
+
+    if (identifier.table_name) {
+      if (identifier.table_name == entry.identifier->table_name &&
+          identifier.column_name == entry.identifier->column_name) {
+        matching_expressions.emplace_back(entry.expression);
+      }
+    } else {
+      if (identifier.column_name == entry.identifier->column_name) {
+        matching_expressions.emplace_back(entry.expression);
+      }
+    }
+  }
+
+  if (matching_expressions.size() != 1) return nullptr; // Identifier is ambiguous/not existing
+
+  return matching_expressions[0];
+}
+
+std::shared_ptr<AbstractExpression> SQLIdentifierContext::resolve_identifier_strict(const SQLIdentifier& identifier) const {
+  const auto expression = resolve_identifier_relaxed(identifier);
+  Assert(expression, "Couldn't resolve identifier "s + identifier.as_string());
+  return expression;
+}
+
+const std::optional<SQLIdentifier> SQLIdentifierContext::get_expression_identifier(const std::shared_ptr<AbstractExpression>& expression) const {
+  auto entry_iter = std::find_if(_entries.begin(), _entries.end(), [&](const auto& entry) { return entry.expression->deep_equals(*expression); });
+  if (entry_iter == _entries.end()) return std::nullopt;
+  return entry_iter->identifier;
+}
+
+std::vector<std::shared_ptr<AbstractExpression>> SQLIdentifierContext::resolve_table_name(const std::string& table_name) const {
```

Comment:
I find it a little bit confusing that I resolve an (unique) identifier like table name and get back a list of expressions. Doesn't this function more search for table name related expressions?

Reply:
Added the comment

```c++
  /**
   * @return   The column expressions of a table/subselect identified by @param table_name.
   */
  std::vector<std::shared_ptr<AbstractExpression>> resolve_table_name(const std::string& table_name) const;
```